R/utils_render_rtf_redo.R

```diff
@@ -240,20 +240,181 @@ rtf_table <- function(rows) {
   rtf_raw(paste(unlist(rows), collapse = ""))
 }
 
+# Conversion factors from absolute width units to twips (`tw`)
+twip_factors <-
+  c(
+    `in` = 1440, `pt` = 20, `px` = 15,
+    `cm` = 566.9291, `mm` = 56.69291, `tw` = 1
+  )
+
+# Input: character vector of any length. Valid values are numbers with
+# suffix of in, pt, px, cm, mm, tw; you can also include `""`.
+# Output: data frame with columns `value` and `unit`, with NA for both
+# if input element was `""`. Values that cannot be parsed will throw errors.
+parse_length_str <- function(lengths_vec,
+                             allow_negative = FALSE) {
+
+  match <- regexec("^([0-9.-]+)(%|[a-z]+)$", lengths_vec)
+  match <- regmatches(lengths_vec, match)
+
+  vals <- vapply(match, function(m) {
+    if (length(m)) {
+      as.numeric(m[[2]])
+    } else {
+      NA_real_
+    }
+  }, numeric(1))
+
+  units <- vapply(match, function(m) {
+    if (length(m)) {
+      m[[3]]
+    } else {
+      NA_character_
+    }
+  }, character(1))
+
+  # Check for negative values and stop of `allow_negative = FALSE`
+  non_na_vals <- vals[!is.na(vals)]
+  if (!allow_negative &&
+      length(non_na_vals) > 0 &&
+      any(!is.na(non_na_vals)) &&
+      is.numeric(non_na_vals) &&
+      any(non_na_vals < 0)) {
+
+    stop("Negative values supplied to widths cannot be used.",
+         call. = FALSE)
+  }
+
+  # Check for bad values and stop if necessary
+  bad_values <- nzchar(lengths_vec) & is.na(vals)
+  if (any(bad_values)) {
+
+    bad_values <- lengths_vec[nzchar(lengths_vec) & is.na(vals)]
+
+    stop(
+      "Some of the values supplied cannot be interpreted:\n",
+      "* Problem values are: ",
+      str_catalog(bad_values, surround = c("\"")), "\n",
+      "* Use either of: `px`, `pt`, `in`, `cm`, `mm`, or `tw` ",
+      "(e.g., \"12px\")",
+      call. = FALSE
+    )
+  }
+
+  dplyr::tibble(value = vals, unit = units)
+}
+
+abs_len_to_twips <- function(lengths_df) {
+
+  lengths_df %>%
+    dplyr::left_join(
+      tibble::enframe(twip_factors, name = "unit", value = "conv"),
+      by = c("unit" = "unit")
+    ) %>%
+    dplyr::mutate(
+      value = ifelse(!is.na(conv), value * conv, value),
+      unit = ifelse(!is.na(conv), "tw", unit)
+    ) %>%
+    dplyr::select(value, unit)
+}
+
+# The `col_width_resolver_rtf()` function returns a vector of
+# RTF column widths in units of twips (numeric)
+#
+# `table_width` and `col_widths` should be character vectors with
+# a numeric portion followed by the unit; the first should be a
+# length-1 vector and the latter should be a vector with as many
+# elements as there are visible columns in the gt table
+col_width_resolver_rtf <- function(table_width,
+                                   col_widths,
+                                   n_cols) {
```

Comment:
I don't love that `n_cols` can potentially be different than `length(col_widths)`. The only reason we need `n_cols` is for the case where `col_widths` is NULL; we want to generate evenly-spaced columns.

Reply:
Also, `col_widths = NULL` should be the default

Reply:
At the very least, we should enforce that if `col_widths` is non-`NULL`, then `n_cols` must be the same as `length(col_widths)` (and we can even make `n_cols = length(col_widths)` the default).

Reply:
The above suggestions were incorporated in `col_width_resolver_rtf()`.